[Merged by Bors] - feat(data/nat/factorization): Defining `factorization`

stuart-presnell opened the pull request:
Defining `factorization` as a `finsupp`, as discussed in
https://leanprover.zulipchat.com/#narrow/stream/113489-new-members/topic/Prime.20factorizations
and 
https://leanprover.zulipchat.com/#narrow/stream/113489-new-members/topic/Proof.20of.20Euler's.20product.20formula.20for.20totient

This is the first of a series of PRs to build up the infrastructure needed for the proof of Euler's product formula for the totient function.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

- [x] depends on: #10554
- [x] depends on: #10880 [using `list.count_singleton'`]

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

stuart-presnell committed:
feat(data/nat/prime_factorization): Defining `prime_factorization`

stuart-presnell committed:
Remove import of `tactic.omega`

stuart-presnell committed:
Add line break to keep lines <100

stuart-presnell committed:
Add doc-string to `prime_factorization`

ericrbg commented on src/data/nat/prime_factorization.lean:
```diff
@@ -0,0 +1,78 @@
+    refine eq_of_count_factors_eq ha hb (λ p, _),
+    simp only [←prime_factorization_count, h] },
+  { intros h, rw h },
+end
```
```suggestion
⟨λ h, eq_of_count_factors_eq ha hb (λ p, by simp [←prime_factorization_count, h]), λ h, by rw h⟩
```

🏌️

stuart-presnell replied:
Very nice, thanks!

ericrbg commented on src/data/nat/prime_factorization.lean:
```diff
@@ -0,0 +1,78 @@
+end
+
+lemma prime_factorization_zero : prime_factorization 0 = 0  :=
+by { unfold prime_factorization, rw factors_zero, simp }
```
`simp` will unfold, and stuff like `factors_zero` are `simp` lemmas; you can replace a lot of these by blocks with just `by simp [prime_factorization]`.

stuart-presnell replied:
Thanks!

ericrbg commented on src/data/nat/prime_factorization.lean:
```diff
@@ -0,0 +1,78 @@
+  exact (factors_eq_nil n),
+end
+
+/-- For prime `p` and `k > 0`, the only prime factor of `p^k` is `p` with multiplicity `k` -/
```
you don't need `k ≠ 0`.

stuart-presnell replied:
Good point, thanks

ericrbg commented on src/data/nat/prime_factorization.lean:
```diff
@@ -0,0 +1,78 @@
+end
+
+/-- The only prime factor of prime `p` is `p` itself, with multiplicity 1 -/
+lemma prime_factorization_prime {p : ℕ} (hp : prime p) : p.prime_factorization = single p 1 :=
```
all of these basic `prime_factorization_{number_type}` lemmas should probably be `simp` lemmas

ericrbg commented on src/data/nat/prime_factorization.lean:
```diff
@@ -0,0 +1,78 @@
+by { unfold prime_factorization, rw factors_one, simp }
+
+/-- The support of `n.prime_factorization` is exactly `n.factors.to_finset` -/
+lemma support_prime_factorization {n : ℕ} : n.prime_factorization.support = n.factors.to_finset :=
```
maybe this one should also be a `simp` lemma too

stuart-presnell committed:
Simpler proofs, per referee suggestion

stuart-presnell committed:
Labelling some lemmas with `@[simp] `

stuart-presnell committed:
Further simplification of proofs

stuart-presnell committed:
Removed 0 < k from `prime_factorization_prime_pos_pow`

eric-wieser commented on src/data/nat/prime_factorization.lean:
```diff
@@ -0,0 +1,72 @@
+noncomputable def prime_factorization (n : ℕ) : ℕ →₀ ℕ := (n.factors : multiset ℕ).to_finsupp
+
+lemma prime_factorization_count {n p : ℕ} : n.prime_factorization p = list.count p n.factors :=
+by simp [prime_factorization]
```
is this true by `rfl`?

stuart-presnell replied:
Apparently not

eric-wieser commented on src/data/nat/prime_factorization.lean:
```diff
@@ -0,0 +1,72 @@
+  split,
+  { simp only [support_prime_factorization, prime_pow_prime_divisor hk hp, finset.subset.refl] },
+  { rw [prime_factorization_count, prime.factors_pow hp k, count_repeat p k] }
+end
```
This lemma looks like a less general version of `pnat.factor_multiset_pow`; can it be stated in the same generality?

stuart-presnell replied:
Yes, thanks for spotting this.  I've added the more general lemma (which is proved via a new addition to `data/nat/prime`, PR #10554) and then rewritten the subsequent lemmas `prime_factorization_prime` and `prime_factorization_prime_pow`.

stuart-presnell committed:
Renamed lemma to `prime_factorization_prime_pow`

stuart-presnell committed:
More general lemma `prime_factorization_pow`

with lemma `factors_count_pow` to be moved into `data/nat/prime`

stuart-presnell committed:
feat(data/nat/prime): Add lemma `factors_count_pow`

stuart-presnell committed:
Moved `factors_count_pow` to separate PR #10554

stuart-presnell committed:
Merge branch 'SP_factors_count_pow' into SP_prime_factorization_basics
Adding dependency on lemma `factors_count_pow`

stuart-presnell committed:
Merge branch 'master' into SP_prime_factorization_basics

YaelDillies commented on src/data/nat/prime_factorization.lean:
```diff
@@ -0,0 +1,82 @@
+ mapping each prime factor of `n` to its multiplicity in `n`. -/
+noncomputable def prime_factorization (n : ℕ) : ℕ →₀ ℕ := (n.factors : multiset ℕ).to_finsupp
+
+lemma prime_factorization_count {n p : ℕ} : n.prime_factorization p = list.count p n.factors :=
```
```suggestion
lemma prime_factorization_eq_count {n p : ℕ} : n.prime_factorization p = n.factors.count p :=
```

YaelDillies commented on src/data/nat/prime_factorization.lean:
```diff
@@ -0,0 +1,82 @@
+by simp [prime_factorization]
+
+/-- Every positive natural number has a unique prime factorization -/
+lemma prime_factorization_eq_iff {a b : ℕ} (ha : 0 < a) (hb : 0 < b) :
```
This name could use `inj` somewhere. And you can state it using `set.inj_on` if you want to, which is quite practical because you get the `↔`, and `a ≠ b → a.prime_factorization ≠ b.prime_factorization` versions for free.
Also, you should ask for `≠ 0`, not `0 <`.

stuart-presnell replied:
Thanks, is the new lemma `factorization_inj` what you were looking for here?

YaelDillies replied:
Which new lemma? 🤔

stuart-presnell replied:
Oh, sorry, I piled up a stack of commits and commented before pushing them!

YaelDillies replied:
Yes it is!

YaelDillies commented on src/data/nat/prime_factorization.lean:
```diff
@@ -0,0 +1,82 @@
+
+/-- `n.prime_factorization` is the finitely supported function `ℕ →₀ ℕ`
+ mapping each prime factor of `n` to its multiplicity in `n`. -/
+noncomputable def prime_factorization (n : ℕ) : ℕ →₀ ℕ := (n.factors : multiset ℕ).to_finsupp
```
This is a bit long of a name. Feel free to shorten it to `prime_fact`, `prime_fac` or `factorization` (after all, it's `nat.factors`, not `nat.prime_factors`).

Same goes for the filename.

YaelDillies commented on src/data/nat/prime_factorization.lean:
```diff
@@ -0,0 +1,82 @@
+  * `prime_factorization 2000 5` is 3
+  * `prime_factorization 2000 k` is 0 for all other `k : ℕ`.
+
+-/
```
Just so that it doesn't mess up docgen
```suggestion
* `prime_factorization 2000 2` is `4`
* `prime_factorization 2000 5` is `3`
* `prime_factorization 2000 k` is `0` for all other `k : ℕ`.
-/
```

YaelDillies commented on src/data/nat/prime_factorization.lean:
```diff
@@ -0,0 +1,82 @@
+  simp only [prime_factorization_count, factors_count_pow],
+end
+
+/-- The only prime factor of prime `p` is `p` itself, with multiplicity 1 -/
```
```suggestion
/-- The only prime factor of prime `p` is `p` itself, with multiplicity `1`. -/
```

YaelDillies commented on src/data/nat/prime_factorization.lean:
```diff
@@ -0,0 +1,82 @@
+  (p ∈ n.prime_factorization.support) ↔ (p ∈ n.factors) :=
+by simp only [support_prime_factorization, list.mem_to_finset]
+
+/-- The only numbers with empty prime factorization are 0 and 1 -/
```
```suggestion
/-- The only numbers with empty prime factorization are `0` and `1`. -/
```

YaelDillies commented on src/data/nat/prime_factorization.lean:
```diff
@@ -0,0 +1,82 @@
+end
+
+/-- The only prime factor of prime `p` is `p` itself, with multiplicity 1 -/
+@[simp] lemma prime_factorization_prime {p : ℕ} (hp : prime p) :
```
You can use dot notation here.
```suggestion
@[simp] lemma prime.prime_factorization {p : ℕ} (hp : prime p) :
```

YaelDillies commented on src/data/nat/prime_factorization.lean:
```diff
@@ -0,0 +1,82 @@
+end
+
+/-- For prime `p` the only prime factor of `p^k` is `p` with multiplicity `k` -/
+@[simp] lemma prime_factorization_prime_pow {p k : ℕ} (hp : prime p) :
```
Same here
```suggestion
@[simp] lemma prime.prime_factorization_pow {p k : ℕ} (hp : prime p) :
```

YaelDillies reviewed (commented):
I'm not sure it's useful given that we already have `multiplicity` and `padic_val_nat`, but looks good!
One thing you could do instead is making `padic_val_nat` a `finsupp`.

You're missing `prime_factorization_mul`.

stuart-presnell committed:
Rename `prime_factorization_count` to `prime_factorization_eq_count`

stuart-presnell committed:
Comment formatting; rename some lemmas

stuart-presnell committed:
Change `'prime_factorization` to `factorization` throughout

stuart-presnell committed:
Create factorization.lean

stuart-presnell committed:
Delete prime_factorization.lean

stuart-presnell committed:
Replace `factorization_eq_iff` with `factorization_inj`

YaelDillies commented on src/data/nat/prime_factorization.lean:
```diff
@@ -0,0 +1,82 @@
+by simp only [support_prime_factorization, list.mem_to_finset]
+
+/-- The only numbers with empty prime factorization are 0 and 1 -/
+lemma prime_factorization_eq_nil_iff (n : ℕ) : n.prime_factorization = 0 ↔ n = 0 ∨ n = 1 :=
```
`nil` is for lists
```suggestion
lemma prime_factorization_eq_zero_iff (n : ℕ) : n.prime_factorization = 0 ↔ n = 0 ∨ n = 1 :=
```

stuart-presnell committed:
Change `nil` to `zero`

stuart-presnell committed:
Add `count_factors_mul_of_pos` to `data/nat/prime`

stuart-presnell committed:
Add `factorization_mul_of_pos`

stuart-presnell committed:
Add docstring to `factorization_mul_of_pos`

YaelDillies commented on src/data/nat/factorization.lean:
```diff
@@ -52,6 +52,10 @@ by simp only [support_factorization, list.mem_to_finset]
 lemma factorization_eq_zero_iff (n : ℕ) : n.factorization = 0 ↔ n = 0 ∨ n = 1 :=
 by simp [factorization, add_equiv.map_eq_zero_iff, multiset.coe_eq_zero]
 
+@[simp] lemma factorization_mul_of_pos {a b : ℕ} (ha : 0 < a) (hb : 0 < b) :
```
Given how expected this statement is, you can shorten it to `factorization_mul`. Also, do not ask for `0 < a` but `a ≠ 0`.

stuart-presnell replied:
Ok.  Why is `a ≠ 0` preferable to `0 < a`?  Should I make the corresponding change to `perm_factors_mul_of_pos` and `count_factors_mul_of_pos` in `data/nat/prime`?

YaelDillies replied:
That's something we've decided. Lemmas about naturals should take `a ≠ 0` as inputs and `0 < a` as outputs. You can change `data.nat.prime` if you want.

ericrbg replied:
This was basically before `bot_lt`, because `le.ne'` is quite short and the old was something like `nat.pos_of_ne_zero`. Now with `bot_lt`, I don't mind as much, but I guess it's still nice

stuart-presnell committed:
Change `factorization_mul_of_pos` to `factorization_mul`

stuart-presnell committed:
Prepare to refactor thru `list.count_singleton`

stuart-presnell committed:
Add `list.count_singleton'` in temporary section

stuart-presnell committed:
Golf proof of `prime.factorization` using `list.count_singleton'`

stuart-presnell committed:
Merge branch 'master' into SP_prime_factorization_basics
Using `count_singleton'`

stuart-presnell committed:
Delete local duplicate of `list.count_singleton'` after merge

ericrbg commented on src/data/nat/factorization.lean:
```diff
@@ -0,0 +1,86 @@
+ mapping each prime factor of `n` to its multiplicity in `n`. -/
+noncomputable def factorization (n : ℕ) : ℕ →₀ ℕ := (n.factors : multiset ℕ).to_finsupp
+
+lemma factorization_eq_count {n p : ℕ} : n.factorization p = list.count p n.factors :=
```
@eric-wieser - do you know why this isn't `rfl`? even when we use `multiset.count`? `finsupp.to_multiset` seems to be defined using `multiset.count`.

ericrbg replied:
i realised in the interim this is due to decidable instances

ericrbg commented on src/data/nat/factorization.lean:
```diff
@@ -0,0 +1,86 @@
+  ext p,
+  simp only [algebra.id.smul_eq_mul, finsupp.coe_smul, pi.smul_apply],
+  simp only [factorization_eq_count, factors_count_pow],
+end
```
```suggestion
by { ext p, simp [factorization_eq_count, factors_count_pow] }
```

ericrbg commented on src/data/nat/factorization.lean:
```diff
@@ -0,0 +1,86 @@
+  rw [factorization_eq_count, factors_prime hp],
+  by_cases hqp : q = p,
+  { rw hqp, simp },
+  { rw [finsupp.single_eq_of_ne (ne.symm hqp), list.count_singleton', if_neg hqp] },
```
```suggestion
  ext q,
  simp_rw [factorization_eq_count, factors_prime hp, count_singleton', single_apply, @eq_comm _ p],
  convert rfl
```

urgh, here's a golf; but the decidability is wrong again... also annoying how `count_singleton'` and `single_apply` have different orders!

stuart-presnell replied:
Thanks for this.  I've golfed it a little more using `if_congr eq_comm`.

I introduced `count_singleton'` myself just a few days ago to simplify this proof, so it's not currently in use elsewhere.  I could swap the order of arguments in it if you think that would be better.  `count a [b] = ite (a = b) 1 0` seemed like the obvious way to write it, but perhaps not.

ericrbg replied:
I'm not sure, maybe? I agree it's the natural order, and yet I also imagine `single_apply` was written that way on purpose. I don't think it matters too much, honestly.

stuart-presnell replied:
Ok, if I find similar issues arising from other uses of `count_singleton'` in future then I'll take that as evidence that the order should be reversed, but until then I'll leave it.

ericrbg commented on src/data/nat/factorization.lean:
```diff
@@ -0,0 +1,86 @@
+/-- For prime `p` the only prime factor of `p^k` is `p` with multiplicity `k` -/
+@[simp] lemma prime.factorization_pow {p k : ℕ} (hp : prime p) :
+  factorization (p^k) = single p k :=
+by simp [factorization_pow, prime.factorization hp]
```
```suggestion
by simp [factorization_pow, hp.factorization]
```

ericrbg reviewed (approved):
lgtm! left some golfs and stuff but not super relevant

stuart-presnell committed:
Edits per reviewer suggestion

jcommelin commented on src/data/nat/factorization.lean:
```diff
@@ -0,0 +1,80 @@
+
+/-- `n.factorization` is the finitely supported function `ℕ →₀ ℕ`
+ mapping each prime factor of `n` to its multiplicity in `n`. -/
+noncomputable def factorization (n : ℕ) : ℕ →₀ ℕ := (n.factors : multiset ℕ).to_finsupp
```
I think this definition should be generalized to a unique factorization monoid that also has a normalization function.

stuart-presnell replied:
I've added a generalized definition to the end of `ring_theory/unique_factorization_domain`, but for now I've left the `nat` version unchanged.

stuart-presnell committed:
Create `UFM_factorization`

Experimental generalisation to unique factorization monoid with a normalization function.

stuart-presnell committed:
Some simple proofs

stuart-presnell committed:
Add TODO

stuart-presnell committed:
Some more theorems

stuart-presnell committed:
Generalise more lemmas

stuart-presnell committed:
Define `factorization_inj'`

stuart-presnell committed:
Add `factorization_pow`

stuart-presnell committed:
Update UFM_factorization.lean

stuart-presnell committed:
Delete UFM_factorization.lean

stuart-presnell committed:
Add `factorization` as a `finsupp` for UFM with normalization

ericrbg commented on src/ring_theory/unique_factorization_domain.lean:
```diff
@@ -1419,3 +1419,46 @@ begin
+
+@[simp] lemma factorization_one : factorization (1 : α) = 0 := by simp [factorization]
+
+/-- The support of `factorization n` is exactly `n.factors.to_finset` -/
```
docstrign refers to the old nat version :)

stuart-presnell committed:
Fix docstrings

ericrbg reviewed (approved):
lgtm!

stuart-presnell committed:
Formatting

stuart-presnell committed:
Update factorization.lean

b-mehta commented on src/ring_theory/unique_factorization_domain.lean:
```diff
@@ -1419,3 +1419,46 @@ begin
+
+/-- For any `p`, the power of `p` in `x^n` is `n` times the power in `x` -/
+lemma factorization_pow {x : α} {n : ℕ} :
+  factorization (x^n) = n • (factorization x) :=
```
```suggestion
  factorization (x^n) = n • factorization x :=
```

stuart-presnell committed:
Update TODO to mention `data/pnat/factors`

YaelDillies commented on src/data/nat/factorization.lean:
```diff
@@ -15,11 +15,13 @@ import data.nat.mul_ind
   * `factorization 2000 k` is 0 for all other `k : ℕ`.
 
-TODO: Move some material on `factors.count`, `padic_val_nat`, and `multiplicity` to this file,
+TODO: As discussed in this Zulip thread:
```
```suggestion
## TODO

As discussed in this Zulip thread:
```

stuart-presnell committed:
Update src/data/nat/factorization.lean

Co-authored-by: Yaël Dillies <yael.dillies@gmail.com>

b-mehta reviewed (approved):
LGTM!

stuart-presnell committed:
Update src/ring_theory/unique_factorization_domain.lean

Co-authored-by: Bhavik Mehta <bhavikmehta8@gmail.com>

stuart-presnell committed:
Add TODO note re: generalization and deduplication

YaelDillies commented on src/data/nat/factorization.lean:
```diff
@@ -15,16 +15,20 @@ import data.nat.mul_ind
   * `factorization 2000 k` is 0 for all other `k : ℕ`.
 
-## TODO
+## TODO:
```
Typo?
```suggestion
## TODO
```

stuart-presnell committed:
Delete colon after "TODO"

Co-authored-by: Yaël Dillies <yael.dillies@gmail.com>
